base/system/winlogon/sas.c

```diff
@@ -352,35 +344,50 @@ PlayLogonSoundThread(
     if (Info.dwCurrentState != SERVICE_RUNNING)
     {
         WARN("wdmaud has not started!\n");
-        return 0;
+        goto Cleanup;
     }
 
     /* Sound subsystem is running. Play logon sound. */
-    TRACE("Playing logon sound\n");
-    if (!ImpersonateLoggedOnUser((HANDLE)lpParameter))
+    TRACE("Playing %s sound\n", SoundData->IsStartup ? "startup" : "logon");
+    if (!ImpersonateLoggedOnUser(SoundData->UserToken))
     {
         ERR("ImpersonateLoggedOnUser failed (%x)\n", GetLastError());
     }
     else
     {
-        PlaySoundRoutine(IsFirstLogon() ? L"SystemStart" : L"WindowsLogon",
+        PlaySoundRoutine(SoundData->IsStartup ? L"SystemStart" : L"WindowsLogon",
                          TRUE,
                          SND_ALIAS | SND_NODEFAULT);
         RevertToSelf();
     }
+
+Cleanup:
+    HeapFree(GetProcessHeap(), 0, SoundData);
     return 0;
 }
 
 static
 VOID
 PlayLogonSound(
-    IN OUT PWLSESSION Session)
+    _In_ PWLSESSION Session)
 {
+    PLOGON_SOUND_DATA SoundData;
     HANDLE hThread;
 
-    hThread = CreateThread(NULL, 0, PlayLogonSoundThread, (PVOID)Session->UserToken, 0, NULL);
-    if (hThread)
-        CloseHandle(hThread);
+    SoundData = HeapAlloc(GetProcessHeap(), 0, sizeof(LOGON_SOUND_DATA));
+    if (!SoundData)
+        return;
+
+    SoundData->UserToken = Session->UserToken;
+    SoundData->IsStartup = IsFirstLogon(Session);
```

Comment:
I'm not convinced you need to introduce a new "SoundData" structure for this (and introduce a new potential "failure" path via the memory allocation). The Session pointer could be passed to the thread instead, from which the PlayLogonSoundThread could directly read off the `Session->UserToken` and the `IsFirstLogon` thing.

(Of course, this requires that the memory buffer containing the session data doesn't get freed while the thread is running.)

Reply:
> The Session pointer could be passed to the thread instead

That's true, but passing `Session` pointer to the thread directly is not safe without introducing a synchronization mechanism (which I think using `LOGON_SOUND_DATA` is simpler and less error prone).

Reply:
Indeed on windows it's guaranteed that handles are multiple of 4

Reply:
I think the allocation is not that bad, considering it's failure path doesn't affect any other function and this code is not performance-critical. Using `LOGON_SOUND_DATA` is also easier to understand the code than inserting a special bit into `HANDLE`.